idaes/apps/caprese/estimator.py

```diff
@@ -0,0 +1,597 @@
+
+        diff_id_list = [id(var[0]) for var in self.differential_vars]
+        for var, val in model_disturbance_weights:  
+            #Check whether the given variable is classified under diffvar
+            if id(var) not in diff_id_list:
+                raise RuntimeError(var.name, 
+                                   " is not a differential variable.")
+
+            if givenform == "weight":
+                self.diffvar_map_moddis[var].weight = val
+            elif givenform == "variance":
+                self.diffvar_map_moddis[var].weight = 1./val
+
+        mea_id_list = [id(var[0]) for var in self.measurement_vars]
+        for var, val in measurement_noise_weights:
+            #Check whether the given variable is declared as a measurement before
+            if id(var) not in mea_id_list:
+                raise RuntimeError(var.name, 
+                                   " is not declared as a measurement.")
+
+            if givenform == "weight":
+                self.meavar_map_meaerr[var].weight = val
+            elif givenform == "variance":
+                self.meavar_map_meaerr[var].weight = 1./val
+
+        moddis_block = self.MODELDISTURBANCE_BLOCK
+        moddis_list = [moddis_block[ind].var for ind in self.DIFFERENTIAL_SET]
+
+        meaerr_block = self.MEASUREMENTERROR_BLOCK
+        meaerr_list = [meaerr_block[ind].var for ind in self.MEASUREMENT_SET]
+
+        wQw = sum(
+            moddis.weight * moddis[sampt]**2
+            for moddis in moddis_list
+            for sampt in self.sample_points
+            if sampt != self.time.first() #Skip model disturbnace at time.first()
+            )
+
+        vRv = sum(
+            meaerr.weight * meaerr[sampt]**2
+            for meaerr in meaerr_list
+            for sampt in self.sample_points #Here should include the time.first()
+            )
+
+        self.noise_minimize_objective = Objective(expr = (wQw) + (vRv))  
+
+    def check_var_con_dof(self, skip_dof_check = False):
+        '''This function does the final check for fixed and unfixed variables
+        as well as the deactivated differential equations. Then check the degree of freedom.
+
+        Parameters
+        ----------
+        skip_dof_check: SKip to check the degree of freedom if it's True. 
+                        Default is False.
+        '''
+
+        self.vectors.input[...].fix()
+        self.vectors.differential[...].unfix()
+        self.vectors.modeldisturbance[:,0].fix(0.0)
+
+        for diff_eq in self.con_category_dict[CC.DIFFERENTIAL]:
+            diff_eq.deactivate()
+
+        n_diffvars = len(self.DIFFERENTIAL_SET.ordered_data())
+        n_steps_in_horizon = len(self.sample_points) - 1
+        correct_dof = n_diffvars * (n_steps_in_horizon + 1)
+
+        if not skip_dof_check:
+            dof = degrees_of_freedom(self)
+            assert dof == correct_dof  
+
+    def load_inputs_for_MHE(self, inputs):
```

Comment:
We should probably have a general `set_inputs_at_time` function that the plant and estimator (and controller if we want) can call in methods like these.

Reply:
yeah, I comment out `load_inputs_for_MHE` and modify `inject_inputs` in dynamic_block. However, I'm struggle with a better way to define the time points in the last sampling time for MHE. Any suggestions?

Reply:
Just pushed a change that calls `self.inject_inputs` from `load_inputs_for_MHE` to save some code.

Reply:
Thanks! I saw them and they look good!